client_api/api/server.go

```diff
@@ -0,0 +1,799 @@
+
+	"crypto/rand"
+	"github.com/docker/notary"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/client/changelist"
+	"github.com/docker/notary/cryptoservice"
+	"github.com/docker/notary/storage"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/trustpinning"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/utils"
+)
+
+// NewServer creates a new instance of a Client API server with a configured
+// upstream Notary Server.
+func NewServer(upstream string, upstreamCAPath string, grpcSrv *grpc.Server) error {
+	srv := &Server{
+		upstream:       upstream,
+		upstreamCAPath: upstreamCAPath,
+	}
+	RegisterNotaryServer(grpcSrv, srv)
+	return nil
+}
+
+type Server struct {
+	upstream       string
+	upstreamCAPath string
+}
+
+func (srv *Server) Initialize(ctx context.Context, initMessage *InitMessage) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(initMessage.Gun))
+	if err != nil {
+		return nil, err
+	}
+
+	roles := make([]data.RoleName, len(initMessage.ServerManagedRoles.Roles))
+	for index, role := range initMessage.ServerManagedRoles.Roles {
+		roles[index] = data.RoleName(role)
+	}
+
+	err = r.Initialize(initMessage.RootKeyIDs, roles...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) Publish(ctx context.Context, gun *GunMessage) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(gun.Gun))
+	if err != nil {
+		return nil, err
+	}
+
+	err = r.Publish()
+	if err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) AddTarget(ctx context.Context, t *Target) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(t.GetGun()))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.AddTarget(
+		&client.Target{
+			Name:   t.GetName(),
+			Hashes: data.Hashes(t.Hashes),
+			Length: t.Length,
+		},
+	); err != nil {
+		return nil, err
+	}
+	if err := publishRepo(r); err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) RemoveTarget(ctx context.Context, t *Target) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(t.GetGun()))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.RemoveTarget(
+		t.GetName(), "targets",
+	); err != nil {
+		return nil, err
+	}
+	if err := publishRepo(r); err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) ListTargets(ctx context.Context, message *RoleNameListMessage) (*TargetWithRoleNameListResponse, error) {
+	r, err := srv.initRepo(data.GUN(message.Gun))
+	if err != nil {
+		return nil, err
+	}
+
+	roles := make([]data.RoleName, len(message.Roles))
+	for index, role := range message.Roles {
+		roles[index] = data.RoleName(role)
+	}
+
+	targets, err := r.ListTargets(roles...)
+	if err != nil {
+		return nil, err
+	}
+
+	resTargets := make([]*TargetWithRole, len(targets))
```

Comment:
Can these be factored out? Maybe something like `NewTargetWithRole(Target)` and `NewTargetWithRoles([]Target)`

They only really need to be serialized here so I understand not separating it out, but if they were separate it would be *very* clear that all this server does is proxy commands to a repo object and then serialize the results.

(Same comment for all of the other serialization happening in this file)